internal/graph/check.go

```diff
@@ -372,28 +372,40 @@ func (c *LocalChecker) checkDirect(parentctx context.Context, req *ResolveCheckR
 		storeID := req.GetStoreID()
 		tk := req.GetTupleKey()
 
+		var allowedTypesForUser []*openfgapb.RelationReference
+		if typesys.GetSchemaVersion() == typesystem.SchemaVersion1_1 {
+			// allowedTypesForUser could be "user" or "user:*" or "group#member"
+			allowedTypesForUser, _ = typesys.GetDirectlyRelatedUserTypes(tuple.GetType(tk.GetObject()), tk.Relation)
+		}
+
 		fn1 := func(ctx context.Context) (*openfgapb.CheckResponse, error) {
-			t, err := c.ds.ReadUserTuple(ctx, storeID, tk)
-			if err != nil {
-				if errors.Is(err, storage.ErrNotFound) {
-					return &openfgapb.CheckResponse{Allowed: false}, nil
-				}
+			if c.shouldCheckDirectTuple(typesys, allowedTypesForUser, tk.GetUser()) {
+				t, err := c.ds.ReadUserTuple(ctx, storeID, tk)
```

Comment:
Instead of nesting this inside the handler, avoid the CheckFuncHandler altogether if there is no need to check it. For example,

```
var handlers []CheckHandlerFunc

var condition bool

if condition {
    handler := func(ctx context.Context) (*openfgapb.CheckResponse, error) {
        ... // leave code as is
    }
   
    handlers = append(handlers, handler)
}

... 

return union(ctx, c.concurrencyLimit, handlers...)
```